dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,519 @@
+    def mkdir(self, parent, name):
+        data = {
+            "name": name,
+            "mimeType": self.MIME_GOOGLE_APPS_FOLDER,
+            "parents": [parent],
+            "spaces": self.space,
+        }
+        return self.request("POST", "drive/v3/files", json=data).json()
+
+    @only_once
+    def makedirs(self, parent, path):
+        current_path = []
+        for part in path.split("/"):
+            current_path.append(part)
+            try:
+                metadata = self.get_metadata_by_path(parent, part)
+                if not self.metadata_isdir(metadata):
+                    raise GDriveError(
+                        "{} is not a folder".format("/".join(current_path))
+                    )
+            except GDriveResourceNotFound:
+                metadata = self.mkdir(parent, part)
+            parent = metadata["id"]
+        return parent
+
+    def _resumable_upload_initiate(self, parent, filename):
+        response = self.request(
+            "POST",
+            "upload/drive/v3/files",
+            params={"uploadType": "resumable"},
+            json={"name": filename, "space": self.space, "parents": [parent]},
+        )
+        return response.headers["Location"]
+
+    def _resumable_upload_first_request(
+        self, resumable_upload_url, from_file, to_info, file_size
+    ):
+        try:
+            # outside of self.request() because this process
+            # doesn't need it to handle errors and retries,
+            # they are handled in the next "while" loop
+            response = self.session.put(
+                resumable_upload_url,
+                data=from_file,
+                headers={"Content-Length": str(file_size)},
+                timeout=self.TIMEOUT,
+            )
+            return response.status_code in (200, 201)
+        # XXX: which exceptions should be handled here?
+        except ConnectionError:
+            logger.info(
+                "got connection error while uploading '{}/{}', "
+                "will resume".format(self.url, to_info.path),
+                exc_info=True,
+            )
+            return False
+
+    def _resumable_upload_resume(
+        self, resumable_upload_url, from_file, to_info, file_size
+    ):
+        try:
+            # determine the offset
+            response = self.session.put(
+                resumable_upload_url,
+                headers={
+                    "Content-Length": str(0),
+                    "Content-Range": "bytes */{}".format(file_size),
+                },
+                timeout=self.TIMEOUT,
+            )
+            if response.status_code in (200, 201):
+                # file has been already uploaded
+                return True
+            elif response.status_code == 404:
+                # restarting upload from the beginning wouldn't make a
+                # profit, so it is better to notify the user
+                raise GDriveError("resumable upload URL has been expired")
```

Comment:
can we raise a human readable explanation? what does this error mean from a user perspective? just a failed upload?